tenders/common/elf.c

```diff
@@ -384,81 +409,111 @@ void elf_load_mft(const char *file, struct mft **mft, size_t *mft_size)
+             */
+            continue;
         nbytes = pread_in_full(fd_kernel, &nhdr, sizeof nhdr,
                 phdr[ph_i].p_offset);
         if (nbytes < 0)
             goto out_error;
         if (nbytes != sizeof nhdr)
             goto out_invalid;
-        if (strncmp(nhdr.name, SOLO5_NOTE_NAME, sizeof(SOLO5_NOTE_NAME)) != 0)
-            continue; /* Not a Solo5 NOTE, next */
-
-        if (nhdr.type != SOLO5_NOTE_MANIFEST) {
-            warnx("%s: phdr[%u] contains invalid Solo5 NOTE", file, ph_i);
+        if (nhdr.h.n_namesz != sizeof(SOLO5_NOTE_NAME))
+            /*
+             * Not a Solo5-owned NOTE or invalid n_namesz, skip.
+             */
+            continue;
+        if (strncmp(nhdr.n_name, SOLO5_NOTE_NAME, sizeof(SOLO5_NOTE_NAME)) != 0)
+            /*
+             * Not a Solo5-owned NOTE, skip.
+             */
+            continue;
+        if (nhdr.h.n_type != note_type)
+            /*
+             * Not the Solo5 NOTE of note_type we are looking for, skip.
+             */
+            continue;
+        /*
+         * Check note descriptor (content) size is within limits, and
+         * cross-check with p_filesz.
+         */
+        if (nhdr.h.n_descsz < 1 || nhdr.h.n_descsz > max_note_size)
             goto out_invalid;
-        }
-        if (nhdr.descsz > MFT_NOTE_MAX_SIZE) {
-            warnx("%s: phdr[%u] Solo5 NOTE size out of range", file, ph_i);
+        if (phdr[ph_i].p_filesz < sizeof nhdr + nhdr.h.n_descsz)
             goto out_invalid;
-        }
 
-        mft_note_found = true;
+        note_found = true;
         break;
     }
-    if (!mft_note_found) {
-        warnx("%s: No manifest found, not a Solo5 executable?", file);
-        goto out_invalid;
+    if (!note_found) {
+        free(ehdr);
+        free(phdr);
+        return -1;
     }
 
     /*
-     * At this point we have verified that the NOTE at phdr[ph_i] could be a
-     * Solo5 manifest, and it's sizes are within some sane limits. Adjust size
-     * and offset to skip the note header itself and read the full note
-     * contents (the manifest) into dynamically allocated memory.
+     * At this point we have verified that the NOTE at phdr[ph_i] is the Solo5
+     * NOTE with the requested note_type and its file sizes are sane.
+     *
+     * Adjust for alignment requested in (note_align) and read the note
+     * descriptor (content) following the header into dynamically allocated
+     * memory.
      */
-    note_offset = phdr[ph_i].p_offset + offsetof(struct mft_note, m);
-    note_size = phdr[ph_i].p_filesz - offsetof(struct mft_note, m);
-    note = malloc(note_size);
-    if (note == NULL)
+    assert(note_align > 0 && (note_align & (note_align - 1)) == 0);
+    note_offset = (sizeof nhdr + (note_align - 1)) & -note_align;
+    assert(note_offset >= sizeof nhdr);
+    note_pad = note_offset - sizeof nhdr;
+    note_size = nhdr.h.n_descsz - note_pad;
+    assert(note_size > 0);
```

Comment:
I believe this will always be true when `note_size != 0` since `note_size` is a `size_t`, so if it underflows it will be (very) large at this point. I didn't verify that such an underflow is achievable, but it seems reasonable given `nhdr.h.n_desc` can be e.g. `2`, and `note_align` is just required to be a non-zero power of 2?

If `!= 0` is the intention, we should have that instead.

But it looks to me like we might want to do the `nhdr.h.n_desc - note_pad` subtraction in a fashion that checks for underflow, since that would also guarantee that it's smaller than `max_note_size`?

Reply:
Got it. How about:

    assert(note_size != 0 && note_size <= nhdr.n_descsz)

?

Reply:
Perfect!